test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestConfiguration.java

```diff
@@ -0,0 +1,115 @@
+
+        QuarkusComponentTest testAnnotation = testClass.getAnnotation(QuarkusComponentTest.class);
+        if (testAnnotation != null) {
+            Collections.addAll(componentClasses, testAnnotation.value());
+            useDefaultConfigProperties = testAnnotation.useDefaultConfigProperties();
+            addNestedClassesAsComponents = testAnnotation.addNestedClassesAsComponents();
+            configSourceOrdinal = testAnnotation.configSourceOrdinal();
+            Class<? extends AnnotationsTransformer>[] transformers = testAnnotation.annotationsTransformers();
+            if (transformers.length > 0) {
+                for (Class<? extends AnnotationsTransformer> transformerClass : transformers) {
+                    try {
+                        annotationsTransformers.add(transformerClass.getDeclaredConstructor().newInstance());
+                    } catch (Exception e) {
+                        LOG.errorf("Unable to instantiate %s", transformerClass);
+                    }
+                }
+            }
+        }
+        // All fields annotated with @Inject represent component classes
+        Class<?> current = testClass;
+        while (current != null) {
+            for (Field field : current.getDeclaredFields()) {
+                if (field.isAnnotationPresent(Inject.class) && !resolvesToBuiltinBean(field.getType())) {
+                    componentClasses.add(field.getType());
+                }
+            }
+            current = current.getSuperclass();
+        }
+        // All static nested classes declared on the test class are components
+        if (addNestedClassesAsComponents) {
+            for (Class<?> declaredClass : testClass.getDeclaredClasses()) {
+                if (Modifier.isStatic(declaredClass.getModifiers())) {
+                    componentClasses.add(declaredClass);
+                }
+            }
+        }
+
+        List<TestConfigProperty> testConfigProperties = new ArrayList<>();
+        Collections.addAll(testConfigProperties, testClass.getAnnotationsByType(TestConfigProperty.class));
+        if (testMethod != null) {
+            Collections.addAll(testConfigProperties, testMethod.getAnnotationsByType(TestConfigProperty.class));
+        }
+        for (TestConfigProperty testConfigProperty : testConfigProperties) {
+            configProperties.put(testConfigProperty.key(), testConfigProperty.value());
+        }
+
+        return new QuarkusComponentTestConfiguration(Map.copyOf(configProperties), List.copyOf(componentClasses),
+                this.mockConfigurators,
+                useDefaultConfigProperties, addNestedClassesAsComponents, configSourceOrdinal,
+                List.copyOf(annotationsTransformers));
+    }
+
+    private boolean resolvesToBuiltinBean(Class<?> rawType) {
+        return Instance.class.isAssignableFrom(rawType) || Event.class.equals(rawType) || BeanManager.class.equals(rawType);
```

Comment:
I fixed this in my classloading hell PR, which we don't really want to merge as is, so maybe we could fix this in this PR instead :-)

I believe the full set of built-in beans we want to detect here is:

- `Provider`
- `Instance`
- `InjectableInstance`
- `Event`
- `BeanContainer`
- `BeanManager`

Reply:
:+1: 

So we'll need to add `Provider` and `BeanContainer`...

Reply:
I was also thinking `InjectableInstance`, but that's not too big of a deal.

Reply:
`Instance.class.isAssignableFrom(rawType)` matches `InjectableInstance`...

Reply:
Oh, you're right. I actually modified that in my branch to an `equals` call, because I didn't want to match `InstanceImpl` by accident, but that's a corner case.